conan/api/subapi/list.py

```diff
@@ -248,8 +258,67 @@ def __init__(self, pref, binary_config, expected_config, remote=None):
         for r in expected_requires:
             existing = binary_requires.get(r.name)
             if not existing or r != existing:
-                self.deps_diff.setdefault("expected", []).append(repr(r))
-                self.deps_diff.setdefault("existing", []).append(repr(existing))
+                self.deps_diff.setdefault("expected", set()).add(repr(r))
+                self.deps_diff.setdefault("existing", set()).add(repr(existing))
+        expected_requires = {r.name: r for r in expected_requires}
+        for r in binary_requires.values():
+            existing = expected_requires.get(r.name)
+            if not existing or r != existing:
+                self.deps_diff.setdefault("expected", set()).add(repr(existing))
+                self.deps_diff.setdefault("existing", set()).add(repr(r))
+
+        # Build requires
+        self.build_requires_diff = {}
+        binary_build_requires = binary_config.get("build_requires", [])
+        expected_build_requires = expected_config.get("build_requires", [])
+        binary_build_requires = [RecipeReference.loads(r) for r in binary_build_requires]
+        expected_build_requires = [RecipeReference.loads(r) for r in expected_build_requires]
+        binary_build_requires = {r.name: r for r in binary_build_requires}
+        for r in expected_build_requires:
+            existing = binary_build_requires.get(r.name)
+            if not existing or r != existing:
+                self.build_requires_diff.setdefault("expected", set()).add(repr(r))
+                self.build_requires_diff.setdefault("existing", set()).add(repr(existing))
```

Comment:
This pattern is repeated a few times, maybe we want to make a local function for it?